Single-branch support for `but branch apply`

Byron opened the pull request:
Port `but branch apply` over to non-legacy code, make sure it builds in non-legacy mode as well.

### Tasks

* [x] add CI validation that it now exists in non-legacy builds.
* [x] add all the tests for the status quo
* [x] de-'legacy' oplog to be able to use 
* [x] new 'apply' function in but-api
* [x] make existing tests work
* [x] fix CI 💢

### Next PR
* [ ] #11400 



Follow-up on #11236.

Byron committed:
refactor tests for the status-quo of `but branch apply` in preparation for apply impl

Byron committed:
Make `but` tests work in non-legacy mode and the capability to CI

This allows us to validate the non-legacy mode as well.

Byron committed:
Create a mixed-mode non-legacy apply.

In legacy mode, we can use non-legacy code, but not the other way around.
So let's split it on the highest level. Also, let's not use feature toggles
for the same reason, it's now not possible anymore to call into old code
as it's not available anymore.

Byron committed:
Provide a non-legacy version of `branch apply`

While at it, make it clearer from dependencies what's legacy and what not,
similar to the `but` (CLI) crate.

Byron committed:
Create a minimal version of `but-oplog` that is mostly legacy, to enable `but-api`

That way we can demonstrate semantics of the oplog/undo system.

krlvi reviewed (approved):
thanks, for this, LGTM

Byron committed:
Make generate-ts easier to handle

Ideally, it's clear why tests fail and how to fix them.

- add a cache
- focus the test-runs to only the crates that harbor ts type definitions
- remove CI dependencies as the dependents hang if there was no Rust change, which
  causes the definitions job to not run.
